nifty-ssl/src/main/java/com/facebook/nifty/ssl/CryptoUtil.java

```diff
@@ -0,0 +1,149 @@
+     */
+    private static byte[] extract(final byte[] salt, final byte[] inputKeyingMaterial) {
+        return initHmacSha256(salt == null ? NULL_SALT : salt).doFinal(inputKeyingMaterial);
+    }
+
+    /**
+     * The "expand" stage of the extract-then-expand HKDF algorithm. All arguments are required (may not be null).
+     *
+     * @param prk the pseudo-random key generated by the extract() stage.
+     * @param previousRoundResult the output of the previous expand() round. For the first round, it should be a
+     *                            zero-length byte array.
+     * @param info the info parameter to hkdf().
+     * @param counter the counter for the current expand stage.
+     * @return output keying material of length SHA256_OUTPUT_BYTES.
+     */
+    private static byte[] expand(final byte[] prk,
+                                 final byte[] previousRoundResult,
+                                 final byte[] info,
+                                 final byte counter) {
+        Mac mac = initHmacSha256(prk);
+        mac.update(previousRoundResult);
+        mac.update(info);
+        mac.update(counter);
+        return mac.doFinal();
+    }
+
+    /**
+     * The HKDF key-derivation function, as defined in RFC 5869 (see https://tools.ietf.org/html/rfc5869).
+     *
+     * @param inputKeyingMaterial the input keying material to HKDF. May not be null.
+     * @param salt optional salt parameter, may be null. This value does not need to be secret and can be safely
+     *             reused between different calls to hkdf(). See the RFC for recommended practices for salt
+     *             selection.
+     * @param info optional application- and/or context-specific information used to bind the derived key material
+     *             to a particular context or use case. See the RFC for recommended practices for info parameter
+     *             selection.
+     * @param outputLength desired length of the output key, in bytes. May not exceed MAX_HKDF_OUTPUT_LENGTH.
+     * @return a pseudo-random key of outputLength bytes long.
+     * @throws IllegalArgumentException if outputLength > MAX_HKDF_OUTPUT_LENGTH.
+     */
+    public static byte[] hkdf(final byte[] inputKeyingMaterial,
+                              final byte[] salt,
+                              final byte[] info,
+                              final int outputLength) {
+        if (outputLength > MAX_HKDF_OUTPUT_LENGTH) {
+            throw new IllegalArgumentException("Output length too large " + outputLength);
+        }
+
+        byte[] prk = extract(salt, inputKeyingMaterial);
+
+        int numRounds = (int) Math.ceil((double) outputLength / SHA256_OUTPUT_BYTES);
+        byte[] outputData = new byte[outputLength];
+
+        int idx = 0;
+        byte[] current = EMPTY_BYTES;
+        final byte[] realInfo = info == null ? EMPTY_BYTES : info; // handle null info w/o checking at every round
+
+        for (int i = 0; i < numRounds; ++i) {
```

Comment:
I think that won't work because Java's byte is signed. So 0xff is actually -127. Basically if numRounds is >= 0x80 then using a byte won't work.